### PR TITLE
Add Azure Network Security Perimeter (NSP) support

### PR DESCRIPTION
**Pull request by eerhardt**
## Description

Implement NSP resources for securing PaaS services (Storage, Key Vault, Cosmos DB, SQL) within a logical network boundary. NSPs are the PaaS-layer equivalent of VNets/NSGs, grouping resources so they can communicate internally while restricting public access via access rules.

New APIs:
- AddNetworkSecurityPerimeter() - creates an NSP with a default profile
- WithAccessRule() - adds inbound/outbound rules (IP prefixes, subscriptions, FQDNs)
- AssociateWith() - explicitly associates a PaaS resource with an NSP

New types:
- AzureNetworkSecurityPerimeterResource
- AzureNspAccessRule (DTO for rule configuration)
- IAzureNspAssociationTarget (implemented by Storage, KeyVault, CosmosDB, SQL)

Associations use Enforced access mode by default. Generated Bicep targets the 2025-05-01 API version via Azure.Provisioning.Network 1.1.0-beta.2.

## Checklist

- Is this feature complete?
  - [x] Yes
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
- Does the change make any security assumptions or guarantees?
  - [x] Yes
    - If yes, have you done a threat model and had a security review?
      - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to `aspire.dev` issue:
      - [Add docs for Azure Network Security Perimeter (NSP) support (microsoft/aspire.dev#689)](https://github.com/microsoft/aspire.dev/issues/689)

### CODE REVIEW

File: src/Aspire.Hosting.Azure.Network/AzureNetworkSecurityPerimeterExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            });
+
+        // Create a default profile
+        var profileIdentifier = Infrastructure.NormalizeBicepIdentifier($"{nsp.BicepIdentifier}_profile");
+        var profile = new NetworkSecurityPerimeterProfile(profileIdentifier)
+        {
+            Name = "defaultProfile",
+            Parent = nsp,
+        };
+        infra.Add(profile);
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure.Network/AzureNetworkSecurityPerimeterExtensions.cs:185]: Note: I'm not currently publicly modeling a "profile" child resource. I'm not sure it is necessary for most usages.

We can easily add it in the future, and keep the existing easy APIs - the existing APIs will still apply to the "default" profile.

Let me know if you think this is a mistake.

**Review by mitchdenny — COMMENTED** [review]
**Finding 1 — Beta package dependency**: `Azure.Provisioning.Network` is upgraded from stable `1.0.0` to pre-release `1.1.0-beta.2`. We should wait until `1.1.0` ships as stable before merging this, so the existing VNet/NSG/PrivateEndpoint features don't ship on a pre-release dependency.

**Comment by mitchdenny** [comment]
**Finding 2 — Duplicate `id` output in Log Analytics Bicep**: The new `output id string` duplicates the existing `output logAnalyticsWorkspaceId string` (both emit `workspace.Id`). I assume the duplication is intentional for consistency — so that all `IAzureNspAssociationTarget` resources use a standard `_outputs_id` parameter naming convention in the NSP Bicep module. Just confirming that's the intent.

**Comment by mitchdenny** [comment]
**Follow-up: E2E deployment tests**

Can we file a follow-up issue to add E2E deployment tests for the NSP feature? Since this PR is from a fork, we can't add deployment E2E tests here (they require secrets/infrastructure), but it would be good to verify that the generated Bicep actually deploys successfully with real Azure resources — especially the association + access rule combinations.

**Comment by mitchdenny** [comment]
**Finding 3 — Missing test for `AssociateWith` duplicate association name validation**

The duplicate name detection in `AssociateWith` (lines 144-149) is untested. There's a `WithAccessRule_DuplicateName_Throws` test but no corresponding test for associations. Proposed test:

```csharp
[Fact]
public void AssociateWith_DuplicateAssociationName_Throws()
{
    using var builder = TestDistributedApplicationBuilder.Create(DistributedApplicationOperation.Publish);

    var nsp = builder.AddNetworkSecurityPerimeter("my-nsp");
    var storage = builder.AddAzureStorage("storage");

    storage.AssociateWith(nsp);

    var exception = Assert.Throws<ArgumentException>(() => storage.AssociateWith(nsp));

    Assert.Contains("storage-assoc", exception.Message, StringComparison.OrdinalIgnoreCase);
}
```

**Comment by mitchdenny** [comment]
**Naming consistency: `NameOutput` vs `NameOutputReference`**

The NSP resource uses `NameOutputReference` which initially looks inconsistent with the other Network resources in this package (VNet, NSG, NatGateway, PublicIP, PrivateEndpoint, PrivateDnsZone) which all use `NameOutput`.

However, `NameOutputReference` is actually the dominant convention across the rest of the repo — used by **24** resource types (KeyVault, Storage, CosmosDB, SQL, EventHubs, ServiceBus, OpenAI, Redis, Foundry, Search, SignalR, AppConfiguration, PostgreSQL, etc.). The 5 `NameOutput` usages are all confined to the Network package's VNet-family resources, making *them* the outliers.

So the NSP resource has the right name here. We should consider a breaking change to rename `NameOutput` → `NameOutputReference` on the existing Network resources (`AzureVirtualNetworkResource`, `AzureNetworkSecurityGroupResource`, `AzureNatGatewayResource`, `AzurePublicIPAddressResource`, `AzurePrivateEndpointResource`, `AzurePrivateDnsZoneResource`) for consistency — either in this PR or as a follow-up.

**Review by mitchdenny — APPROVED** [review]
This looks good to me. We'll see if we need profile support. I think we might want to make a breaking change re `NameOutput` just for consistency.

**Comment by mitchdenny** [comment]
One more question ... how do we feel about `AssociateWith(...)` as a name. Do we think that this is a common pattern .. or should we do something like `WithNetworkSecurityPerimeter(..)`?

**Comment by eerhardt** [comment]
Thanks for the review @mitchdenny.

> Finding 2 — Duplicate id output in Log Analytics Bicep

I removed the duplicate output in the bicep. I kept the "Id" C# property- this will just have both.

> Finding 1 — Beta package dependency

We will get the Azure.Provisioning team to release stable. I will merge the PR and we can update to the stable version when it is ready. It will be before we branch for 13.3.

> Follow-up: E2E deployment tests

Yep, I will create a PR to add a new deployment test once this is merged.

> Finding 3 — Missing test for AssociateWith duplicate association name validation

Thanks for adding it to this PR.

> Naming consistency: NameOutput vs NameOutputReference

Fixed with [Rename NameOutput to NameOutputReference in Aspire.Hosting.Azure.Network (microsoft/aspire#16004)](https://github.com/microsoft/aspire/pull/16004)[Rename NameOutput to NameOutputReference in Aspire.Hosting.Azure.Network (microsoft/aspire#16004)](https://github.com/microsoft/aspire/pull/16004)

> how do we feel about AssociateWith(...)

Good suggestion, I renamed it to `WithNetworkSecurityPerimeter`.
